cpp/src/io/orc/reader_impl.cu

```diff
@@ -215,6 +218,257 @@ size_t gather_stream_info(const size_t stripe_index,
+    std::vector<std::vector<size_type>> const &user_specified_stripes,
+    size_type &row_start,
+    size_type &row_count)
+  {
+    std::vector<cudf::io::orc::metadata::stripe_source_mapping> selected_stripes_mapping;
+
+    if (!user_specified_stripes.empty()) {
+      CUDF_EXPECTS(user_specified_stripes.size() == get_num_source_files(),
+                   "Must specify stripes for each source");
+      // row_start is 0 if stripes are set. If this is not true anymore, then
+      // row_start needs to be subtracted to get the correct row_count
+      CUDF_EXPECTS(row_start == 0, "Start row index should be 0");
+
+      row_count = 0;
+      // Each vector entry represents a source file; each nested vector represents the
+      // user_defined_stripes to get from that source file
+      for (size_t src_file_idx = 0; src_file_idx < user_specified_stripes.size(); ++src_file_idx) {
+        std::vector<int> stripe_idxs;
+        std::vector<OrcStripeInfo> stripe_infos;
+
+        // Coalesce stripe info at the source file later since that makes downstream processing much
+        // easier in impl::read
+        for (const size_t &stripe_idx : user_specified_stripes[src_file_idx]) {
+          CUDF_EXPECTS(stripe_idx < per_file_metadata[src_file_idx].ff.stripes.size(),
+                       "Invalid stripe index");
+          stripe_idxs.push_back(stripe_idx);
+          stripe_infos.push_back(
+            std::make_pair(&per_file_metadata[src_file_idx].ff.stripes[stripe_idx], nullptr));
+          row_count += per_file_metadata[src_file_idx].ff.stripes[stripe_idx].numberOfRows;
+        }
+        selected_stripes_mapping.push_back(
+          {static_cast<int>(src_file_idx), stripe_idxs, stripe_infos});
+      }
+    } else {
+      row_start = std::max(row_start, 0);
+      if (row_count < 0) {
+        row_count = static_cast<size_type>(
+          std::min<int64_t>(get_num_rows(), std::numeric_limits<size_type>::max()));
+      }
+      row_count = std::min(row_count, get_num_rows() - row_start);
+      CUDF_EXPECTS(row_count >= 0, "Invalid row count");
+      CUDF_EXPECTS(row_start <= get_num_rows(), "Invalid row start");
+
+      size_type count = 0;
+      // Iterate all source files, each source file has corelating metadata
+      for (size_t src_file_idx = 0; src_file_idx < per_file_metadata.size(); ++src_file_idx) {
+        std::vector<int> stripe_idxs;
+        std::vector<OrcStripeInfo> stripe_infos;
+
+        for (size_t stripe_idx = 0; stripe_idx < per_file_metadata[src_file_idx].ff.stripes.size();
+             ++stripe_idx) {
+          count += per_file_metadata[src_file_idx].ff.stripes[stripe_idx].numberOfRows;
+          if (count > row_start || count == 0) {
+            stripe_idxs.push_back(stripe_idx);
+            stripe_infos.push_back(
+              std::make_pair(&per_file_metadata[src_file_idx].ff.stripes[stripe_idx], nullptr));
+          }
+          if (count >= row_start + row_count) { break; }
```

Comment:
You have to break from main for loop as well since count is already met and we don't need to read any more stripes from any other source.